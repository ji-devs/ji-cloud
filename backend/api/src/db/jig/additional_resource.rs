use shared::domain::jig::{AdditionalResourceId, JigId};
use sqlx::PgPool;

pub async fn create(
    pool: &PgPool,
    parent: JigId,
    url: String,
) -> anyhow::Result<AdditionalResourceId> {
    sqlx::query!(
        r#"
insert into jig_additional_resource (jig_id, url)
values ($1, $2)
returning id
        "#,
        parent.0,
        url,
    )
    .fetch_one(pool)
    .await
    .map(|it| AdditionalResourceId(it.id))
    .map_err(Into::into)
}

pub async fn get(
    pool: &PgPool,
    parent: JigId,
    id: AdditionalResourceId,
) -> anyhow::Result<Option<String>> {
    let url = sqlx::query!(
        r#"
select
    url as "url!: String"
from jig_additional_resource
where jig_id = $1 and id = $2
        "#,
        parent.0,
        id.0,
    )
    .fetch_optional(pool)
    .await?
    .map(|it| it.url);

    Ok(url)
}

pub async fn update(
    pool: &PgPool,
    parent_id: JigId,
    id: AdditionalResourceId,
    url: Option<String>,
) -> anyhow::Result<bool> {
    let mut txn = pool.begin().await?;

    sqlx::query!(
        r#"
update jig_additional_resource
set url = coalesce($3, url)
where jig_id = $1 and id = $2
        "#,
        parent_id.0,
        id.0,
        url,
    )
    .execute(&mut txn)
    .await?;

    txn.commit().await?;

    Ok(true)
}

pub async fn delete(
    pool: &PgPool,
    parent_id: JigId,
    id: AdditionalResourceId,
) -> anyhow::Result<()> {
    let mut txn = pool.begin().await?;

    sqlx::query!(
        r#"
delete from jig_additional_resource
where jig_id = $1 and id = $2
        "#,
        parent_id.0,
        id.0,
    )
    .execute(&mut txn)
    .await?;

    txn.commit().await?;

    Ok(())
}
