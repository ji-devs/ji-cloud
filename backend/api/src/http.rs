use std::{net::TcpListener, sync::Arc};

use actix_service::Service;
use actix_web::{
    dev::{MessageBody, Server, ServiceRequest, ServiceResponse},
    error::Error,
    middleware::{Compat, Condition},
    web::{method, Data},
    HttpResponse,
};
use core::{
    config::JSON_BODY_LIMIT,
    env::env_bool,
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use sqlx::postgres::PgPool;
use tracing::Span;
use tracing_actix_web::{root_span, DefaultRootSpanBuilder, RootSpanBuilder, TracingLogger};

use crate::{
    error::BasicError,
    service::{self, mail, s3, upload::cleaner, ServiceData},
    translate,
};

mod cors;
mod endpoints;

fn log_ise<B: MessageBody, T>(
    request: ServiceRequest,
    srv: &T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let uri: serde_json::Value = request.uri().to_string().into();
    let method: serde_json::Value = request.method().to_string().into();

    let future = srv.call(request);
    async {
        let mut result = future.await?;
        if result.status() == 500 {
            let response: &mut actix_web::HttpResponse<_> = result.response_mut();

            if let Some(err) = response.extensions_mut().remove::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
                sentry::add_breadcrumb(sentry::Breadcrumb {
                    ty: "http".to_owned(),
                    category: Some("request".into()),
                    data: {
                        let mut map = sentry::protocol::Map::new();
                        map.insert("url".to_owned(), uri);
                        map.insert("method".to_owned(), method);
                        map
                    },
                    ..Default::default()
                });

                sentry::integrations::anyhow::capture_anyhow(&err);
            }
        }

        Ok(result)
    }
}

pub struct Application {
    port: u16,
    server: Option<Server>,
}

impl Application {
    pub fn new(port: u16, server: Server) -> Self {
        Self {
            port,
            server: Some(server),
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    // A more expressive name that makes it clear that
    // this function only returns when the application is stopped.
    pub async fn run_until_stopped(mut self) -> Result<(), std::io::Error> {
        if let Some(server) = self.server.take() {
            server.await?;
        }

        Ok(())
    }

    pub async fn stop(mut self, graceful: bool) {
        if let Some(server) = self.server.take() {
            server.stop(graceful).await
        }
    }
}

impl Drop for Application {
    fn drop(&mut self) {
        if let Some(server) = self.server.take() {
            let _ = tokio::spawn(server.stop(false));
        }
    }
}

#[actix_web::main]
pub async fn build_and_run(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<()> {
    let app = build(
        pool,
        settings,
        s3,
        gcp_key_store,
        gcs,
        algolia,
        algolia_key_store,
        jwk_verifier,
        mail_client,
        algolia_manager,
        media_upload_cleaner,
        google_translate,
    )?;
    app.run_until_stopped().await?;

    Ok(())
}

pub fn build(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<Application> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;

    let s3 = s3.map(ServiceData::new);
    let gcp_key_store = gcp_key_store.map(ServiceData::new);
    let gcs = gcs.map(ServiceData::new);
    let algolia = algolia.map(ServiceData::new);
    let algolia_key_store = algolia_key_store.map(ServiceData::new);
    let mail_client = mail_client.map(ServiceData::new);
    let algolia_manager = algolia_manager.map(ServiceData::new);
    let media_upload_cleaner = media_upload_cleaner.map(ServiceData::new);
    let google_translate = google_translate.map(ServiceData::new);

    let server = actix_web::HttpServer::new(move || {
        let app = actix_web::App::new()
            .app_data(Data::new(pool.clone()))
            .app_data(Data::new(settings.clone()));

        let app = match s3.clone() {
            Some(s3) => app.app_data(s3),
            None => app,
        };

        let app = match gcp_key_store.clone() {
            Some(gcp_key_store) => app.app_data(gcp_key_store),
            None => app,
        };

        let app = match gcs.clone() {
            Some(gcs) => app.app_data(gcs),
            None => app,
        };

        let app = match algolia.clone() {
            Some(algolia) => app.app_data(algolia),
            None => app,
        };

        let app = match algolia_key_store.clone() {
            Some(algolia_key_store) => app.app_data(algolia_key_store),
            None => app,
        };

        let app = match mail_client.clone() {
            Some(mail_client) => app.app_data(mail_client),
            None => app,
        };

        let app = match algolia_manager.clone() {
            Some(algolia_manager) => app.app_data(algolia_manager),
            None => app,
        };

        let app = match media_upload_cleaner.clone() {
            Some(media_upload_cleaner) => app.app_data(media_upload_cleaner),
            None => app,
        };
        let app = match google_translate.clone() {
            Some(google_translate) => app.app_data(google_translate),
            None => app,
        };

        let enable_tracing_logs = env_bool("ENABLE_TRACING_LOGS");

        app.app_data(Data::from(jwk_verifier.clone()))
            .wrap(cors::get(local_insecure))
            .wrap(Condition::new(
                !enable_tracing_logs,
                Compat::new(actix_web::middleware::Logger::default()),
            ))
            .wrap(Condition::new(
                enable_tracing_logs,
                Compat::new(TracingLogger::<JigziSpanBuilder>::new()),
            ))
            .wrap_fn(log_ise)
            .app_data(
                actix_web::web::JsonConfig::default()
                    .limit(JSON_BODY_LIMIT as usize)
                    .error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::QueryConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::PathConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .external_resource(
                "google_cloud_oauth",
                "https://accounts.google.com/o/oauth2/v2/auth",
            )
            .default_service(actix_web::web::to(default_route))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
            .configure(endpoints::audio::configure)
            .configure(endpoints::meta::configure)
            .configure(endpoints::jig::configure)
            .configure(endpoints::module::configure)
            .configure(endpoints::additional_resource::configure)
            .configure(endpoints::jig::report::configure)
            .configure(endpoints::jig::curation::configure)
            .configure(endpoints::course::configure)
            .configure(endpoints::admin::configure)
            .configure(endpoints::animation::configure)
            .configure(endpoints::search::configure)
            .configure(endpoints::media::configure)
            .configure(endpoints::session::configure)
            .configure(endpoints::locale::configure)
            .configure(endpoints::scheduler::configure)
            .configure(endpoints::pdf::configure)
            .configure(endpoints::badge::configure)
            .route("/", method(http::Method::GET).to(no_content_response))
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let listener = if let Some(l) = get_tcp_fd() {
        l
    } else {
        TcpListener::bind(get_addr(Some(api_port)))?
    };

    let port = listener.local_addr().unwrap().port();

    let server = server.listen(listener)?;

    Ok(Application::new(port, server.run()))
}

fn default_route() -> HttpResponse {
    HttpResponse::NotFound().json(BasicError::with_message(
        http::StatusCode::NOT_FOUND,
        "Route not found".to_owned(),
    ))
}

async fn no_content_response() -> HttpResponse {
    HttpResponse::NoContent().finish()
}

pub fn bad_request_handler() -> actix_web::Error {
    BasicError::new(http::StatusCode::BAD_REQUEST).into()
}

pub struct JigziSpanBuilder;

impl RootSpanBuilder for JigziSpanBuilder {
    fn on_request_start(request: &ServiceRequest) -> Span {
        let route = request.match_pattern().unwrap_or("default".to_string());
        let transaction_name = format!("{} {}", request.method(), route);
        let transaction_name = transaction_name.as_str();
        root_span!(request, message = transaction_name)
    }

    fn on_request_end<B>(span: Span, outcome: &Result<ServiceResponse<B>, Error>) {
        DefaultRootSpanBuilder::on_request_end(span, outcome)
    }
}
