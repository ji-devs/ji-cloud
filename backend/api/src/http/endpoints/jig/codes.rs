use actix_web::{
    web::{self, Data, Json, Query},
    HttpResponse,
};
use shared::{
    api::{endpoints::jig::codes, ApiEndpoint},
    domain::jig::codes::{JigCode, JigCodeListResponse, JigCodeSessionsListResponse},
};
use sqlx::PgPool;

use crate::{db, error, extractor::TokenUser};

/// Create a jig player session for the author, if one does not exist already.
pub async fn create(
    db: Data<PgPool>,
    claims: TokenUser,
    req: Json<<codes::Create as ApiEndpoint>::Req>,
) -> Result<HttpResponse, error::JigCode> {
    let req = req.into_inner();
    let user_id = claims.user_id();

    db::jig::is_logged_in(&*db, user_id).await?;

    let jig_code = db::jig::codes::create(&db, user_id, &req).await?;

    Ok(HttpResponse::Created().json(jig_code))
}

pub async fn update(
    db: Data<PgPool>,
    claims: TokenUser,
    path: web::Path<JigCode>,
    req: Json<<codes::Update as ApiEndpoint>::Req>,
) -> Result<HttpResponse, error::JigCode> {
    let code = path.into_inner();
    let req = req.into_inner();
    let user_id = claims.user_id();

    db::jig::is_users_code(&*db, user_id, code).await?;

    db::jig::codes::update(&db, code, &req).await?;

    Ok(HttpResponse::NoContent().into())
}

/// Get all jig codes for user.
pub async fn list_user_codes(
    db: Data<PgPool>,
    claims: TokenUser,
    query: Option<Query<<codes::JigCodeList as ApiEndpoint>::Req>>,
) -> Result<Json<<codes::JigCodeList as ApiEndpoint>::Res>, error::JigCode> {
    let user_id = claims.user_id();
    let query = query.map_or_else(Default::default, Query::into_inner);

    let codes = db::jig::codes::list_user_codes(&*db, user_id, query).await?;

    Ok(Json(JigCodeListResponse { codes }))
}

/// Fetch all sessions for a code.
pub async fn list_code_sessions(
    db: Data<PgPool>,
    claims: TokenUser,
    path: web::Path<JigCode>,
) -> Result<Json<<codes::JigCodeSessions as ApiEndpoint>::Res>, error::JigCode> {
    let code = path.into_inner();
    let user_id = claims.user_id();

    let sessions = db::jig::codes::list_code_sessions(&*db, user_id, code).await?;

    Ok(Json(JigCodeSessionsListResponse { sessions }))
}

pub mod instance {
    use actix_web::{
        web::{Data, Json},
        HttpResponse,
    };
    use chrono::{Duration, Utc};
    use ji_core::settings::RuntimeSettings;
    use serde::Deserialize;
    use shared::{
        api::{endpoints::jig::codes, ApiEndpoint},
        domain::jig::codes::instance::PlayerSessionInstanceResponse,
    };
    use sqlx::PgPool;

    use crate::{
        db, error,
        extractor::IPAddress,
        token::{create_player_session_instance_token, validate_token},
    };
    use uuid::Uuid;

    /// Create a jig player session instance
    pub async fn start_session(
        settings: Data<RuntimeSettings>,
        db: Data<PgPool>,
        ip_address: IPAddress,
        req: Json<<codes::instance::Create as ApiEndpoint>::Req>,
    ) -> Result<
        (
            Json<<codes::instance::Create as ApiEndpoint>::Res>,
            actix_web::http::StatusCode,
        ),
        error::JigCode,
    > {
        let req = req.into_inner();

        let resp = db::jig::codes::start_session(&*db, req.code, ip_address).await?;

        let token: String = create_player_session_instance_token(
            &settings.token_secret,
            Duration::weeks(2),
            &resp.2,
            Utc::now(),
        )?;

        Ok((
            Json(PlayerSessionInstanceResponse {
                jig_id: resp.0,
                settings: resp.1,
                token,
            }),
            actix_web::http::StatusCode::CREATED,
        ))
    }

    #[derive(Deserialize)]
    struct InstanceToken {
        /// The instance this token is for.
        pub sub: Uuid,
    }

    /// Create a jig player session for someone who's not the author, if one doesn't already exist
    pub async fn complete_session(
        settings: Data<RuntimeSettings>,
        db: Data<PgPool>,
        ip_address: IPAddress,
        req: Json<<codes::instance::Complete as ApiEndpoint>::Req>,
    ) -> Result<HttpResponse, error::JigCode> {
        let req = req.into_inner();

        let token = validate_token(&req.token, None, &settings.token_secret)
            .map_err(|_| error::JigCode::Forbidden)?;

        let instance_token: InstanceToken = serde_json::from_value(token)?;

        db::jig::codes::complete_session(
            &db,
            req.session,
            req.players_name,
            instance_token.sub,
            ip_address,
        )
        .await?;

        Ok(HttpResponse::NoContent().finish())
    }
}
