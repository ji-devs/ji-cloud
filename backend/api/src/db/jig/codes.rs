use chrono::{DateTime, Duration, Utc};
use rand::{rngs::ThreadRng, Rng};
use shared::config::{JIG_PLAYER_SESSION_CODE_MAX, JIG_PLAYER_SESSION_VALID_DURATION_SECS};
use shared::domain::jig::codes::{
    JigCodeListRequest, JigCodeSessionResponse, JigCodeUpdateRequest, JigPlayerSessionCreateRequest,
};
use shared::domain::jig::{
    codes::{JigCode, JigCodeResponse, JigPlaySession},
    player::JigPlayerSettings,
    JigId,
};
use shared::domain::user::UserId;
use sqlx::{error::DatabaseError, postgres::PgDatabaseError, PgPool};
use uuid::Uuid;

use shared::domain::jig::TextDirection;

use crate::error;
use crate::extractor::IPAddress;

pub async fn create(
    db: &PgPool,
    creator_id: UserId,
    opts: &JigPlayerSessionCreateRequest,
) -> Result<JigCodeResponse, error::JigCode> {
    let mut generator = rand::thread_rng();

    let mut code = generate_random_code(&mut generator);

    let expires_at = Utc::now() + Duration::seconds(JIG_PLAYER_SESSION_VALID_DURATION_SECS as i64);

    // retry as many times as there are possible codes
    // NOTE: this is NOT guaranteed to successfully insert if there
    for _ in 0..JIG_PLAYER_SESSION_CODE_MAX * 2 {
        match sqlx::query!(
            //language=SQL
            r#"
insert into jig_code (jig_id, creator_id, name, code, direction, scoring, drag_assist, expires_at)
values ($1, $2, $3, $4, $5, $6, $7, $8)
returning created_at as "created_at: DateTime<Utc>"
"#,
            opts.jig_id.0,
            creator_id.0,
            opts.name,
            code,
            opts.settings.direction as i16,
            opts.settings.scoring,
            opts.settings.drag_assist,
            expires_at,
        )
        .fetch_one(db)
        .await
        {
            Ok(res) => {
                return Ok(JigCodeResponse {
                    index: JigCode(code),
                    jig_id: opts.jig_id,
                    name: opts.name.clone(),
                    settings: opts.settings.clone(),
                    created_at: res.created_at,
                    expires_at,
                })
            }
            Err(err) => match err {
                sqlx::Error::Database(db_err) => {
                    session_create_error_or_continue(db_err)?;
                    // did not return error on previous line, retry with new code
                    code = generate_random_code(&mut generator);
                }
                err => return Err(anyhow::anyhow!("sqlx error: {:?}", err).into()),
            },
        }
    }

    Err(anyhow::anyhow!("Maximum retries reached for creating a new jig session").into())
}

pub async fn update(
    db: &PgPool,
    code: JigCode,
    opts: &JigCodeUpdateRequest,
) -> Result<(), error::JigCode> {
    let name = opts.name.clone();
    let direction = opts.settings.as_ref().map(|opts| opts.direction);
    let scoring = opts.settings.as_ref().map(|opts| opts.scoring);
    let drag_assist = opts.settings.as_ref().map(|opts| opts.drag_assist);

    sqlx::query!(
        //language=SQL
        r#"
            update jig_code
            set name = case when $2 then $3 else name end,
                direction = coalesce($4, direction),
                scoring = coalesce($5, scoring),
                drag_assist = coalesce($6, drag_assist)
            where code = $1
        "#,
        code.0,
        name.is_some(),
        name.flatten(),
        direction.map(|d| d as i16),
        scoring,
        drag_assist,
    )
    .execute(db)
    .await?;

    Ok(())
}

fn session_create_error_or_continue(db_err: Box<dyn DatabaseError>) -> Result<(), error::JigCode> {
    let constraint = db_err.downcast_ref::<PgDatabaseError>().constraint();

    match constraint {
        Some("jig_player_session_pkey") => {
            // same code but different jig. retry insert with a new code
            Ok(())
        }
        Some("jig_player_session_jig_id_fkey") => {
            // no jig with this id exists
            Err(error::JigCode::ResourceNotFound)
        }
        db_err => Err(anyhow::anyhow!("{}", db_err.unwrap_or("unknown database error")).into()),
    }
}

fn generate_random_code(generator: &mut ThreadRng) -> i32 {
    debug_assert!(JIG_PLAYER_SESSION_CODE_MAX > 0);

    generator.gen_range(0..JIG_PLAYER_SESSION_CODE_MAX)
}

pub async fn get_code(db: &PgPool, code: JigCode) -> sqlx::Result<JigCodeResponse> {
    let row = sqlx::query!(
        //language=SQL
        r#"
            select code as "code!: i32",
                jig_id as "jig_id: JigId",
                direction as "direction: TextDirection",
                scoring,
                drag_assist,
                name as "name?",
                created_at as "created_at: DateTime<Utc>",
                expires_at as "expires_at: DateTime<Utc>"
            from jig_code
            where code = $1
        "#,
        code.0,
    )
    .fetch_one(db)
    .await?;

    let response = JigCodeResponse {
        index: JigCode(row.code),
        jig_id: row.jig_id,
        name: row.name,
        settings: JigPlayerSettings {
            direction: row.direction,
            scoring: row.scoring,
            drag_assist: row.drag_assist,
        },
        created_at: row.created_at,
        expires_at: row.expires_at,
    };

    Ok(response)
}

pub async fn list_user_codes(
    db: &PgPool,
    user_id: UserId,
    query: JigCodeListRequest,
) -> sqlx::Result<Vec<JigCodeResponse>> {
    let sessions = sqlx::query!(
        //language=SQL
        r#"
select code     as "code!: i32",
       jig_id as "jig_id: JigId",
       direction as "direction: TextDirection",
       scoring,
       drag_assist,
       name as "name?",
       created_at as "created_at: DateTime<Utc>",
       expires_at as "expires_at: DateTime<Utc>"
from jig_code
where creator_id = $1 AND (jig_id = $2 or $2 is null)
"#,
        user_id.0,
        query.jig_id.map(|j| j.0),
    )
    .fetch_all(db)
    .await?
    .into_iter()
    .map(|it| JigCodeResponse {
        index: JigCode(it.code),
        jig_id: it.jig_id,
        name: it.name,
        settings: JigPlayerSettings {
            direction: it.direction,
            scoring: it.scoring,
            drag_assist: it.drag_assist,
        },
        created_at: it.created_at,
        expires_at: it.expires_at,
    })
    .collect();

    Ok(sessions)
}

pub async fn list_code_sessions(
    db: &PgPool,
    user_id: UserId,
    code: JigCode,
) -> Result<Vec<JigCodeSessionResponse>, error::JigCode> {
    // ensure this user's code
    sqlx::query!(
        //language=SQL
        r#"
            SELECT * FROM jig_code
            WHERE creator_id = $1 AND code = $2;
        "#,
        user_id.0,
        code.0
    )
    .fetch_one(db)
    .await?;

    let sessions = sqlx::query!(
        //language=SQL
        r#"
            SELECT
                id,
                code,
                players_name,
                started_at,
                finished_at,
                info
            FROM jig_code_session
            WHERE code = $1 AND finished_at IS NOT NULL
            ORDER BY started_at;
        "#,
        code.0
    )
    .fetch_all(db)
    .await?
    .into_iter()
    .map(|it| {
        Ok(JigCodeSessionResponse {
            code: JigCode(it.code),
            players_name: it.players_name,
            started_at: it.started_at,
            finished_at: it.finished_at,
            info: match it.info {
                Some(r) => serde_json::from_value(r)?,
                None => None,
            },
        })
    })
    .collect::<Result<Vec<_>, error::JigCode>>()?;

    Ok(sessions)
}

/// Creates new jig player session for a player
pub async fn start_session(
    db: &PgPool,
    code: JigCode,
    ip_address: IPAddress,
) -> Result<(JigId, JigPlayerSettings, Uuid), error::JigCode> {
    let mut txn = db.begin().await?;

    let session_info = sqlx::query!(
        //language=SQL
        r#"
        select jig_id as "jig_id: JigId", 
               direction as "direction: TextDirection", 
               scoring,
               drag_assist
        from jig_code
        where code=$1
        "#,
        code.0
    )
    .fetch_optional(&mut txn)
    .await?
    .ok_or(error::JigCode::ResourceNotFound)?;

    // insert into the jig_code_session table returning the instance_id
    let instance_id = sqlx::query!(
        //language=SQL
        r#"
        insert into jig_code_session (code, started_at, ip_address)
        values ($1, current_timestamp, $2)
        returning id as "id: Uuid"
        "#,
        code.0,
        ip_address.0,
    )
    .fetch_one(&mut txn)
    .await?
    .id;

    txn.commit().await?;

    Ok((
        session_info.jig_id,
        JigPlayerSettings {
            direction: session_info.direction,
            scoring: session_info.scoring,
            drag_assist: session_info.drag_assist,
        },
        instance_id,
    ))
}

/// Completes a jig player session for a player and updates play count
pub async fn complete_session(
    db: &PgPool,
    session: JigPlaySession,
    players_name: Option<String>,
    instance_id: Uuid,
    ip_address: IPAddress,
) -> Result<(), error::JigCode> {
    let session = serde_json::to_value(&session)?;
    sqlx::query!(
        //language=SQL
        r#"
            UPDATE jig_code_session
            SET finished_at = current_timestamp, info=$1, players_name=$2
            WHERE id = $3 and ip_address = $4 and finished_at is null;
        "#,
        session,
        players_name,
        instance_id,
        ip_address.0,
    )
    .execute(db)
    .await?;

    Ok(())
}
