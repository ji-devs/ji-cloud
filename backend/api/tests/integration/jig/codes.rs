use crate::{
    fixture::Fixture,
    helpers::{setup_service, LoginExt},
};
use http::StatusCode;
use macros::test_service;
use shared::domain::jig::codes::{
    instance::PlayerSessionInstanceResponse, JigCodeListResponse, JigCodeResponse,
};
use sqlx::postgres::{PgConnectOptions, PgPoolOptions};

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::MetaKinds", "Fixture::User", "Fixture::Jig")
)]
async fn create_and_list(port: u16) -> anyhow::Result<()> {
    let name = "create";

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig/codes", port))
        .json(&serde_json::json!({
            "jigId": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
            "settings": {
                "direction": "rtl",
                "scoring": false,
                "drag_assist": false,
            }
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: JigCodeResponse = resp.json().await?;

    insta::assert_json_snapshot!(format!("{}-1",name), body, { ".**.index" => "[index]", ".**.created_at" => "[timestamp]", ".**.expires_at" => "[timestamp]" });

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/jig/codes/{}",
            port,
            body.index.to_string()
        ))
        .json(&serde_json::json!({
            "name": "test-name"
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let _resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig/codes", port))
        .json(&serde_json::json!({
            "jigId": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
            "settings": {
                "direction": "rtl",
                "scoring": true,
                "drag_assist": false,
            }
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/jig/codes", port))
        .json(&serde_json::json!({}))
        .login()
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: JigCodeListResponse = resp.json().await?;

    insta::assert_json_snapshot!(format!("{}-2",name), body, { ".**.index" => "[index]", ".**.created_at" => "[timestamp]", ".**.expires_at" => "[timestamp]" });

    Ok(())
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::MetaKinds", "Fixture::User", "Fixture::Jig")
)]
async fn session_instance_play_count_flow(port: u16) -> anyhow::Result<()> {
    let name = "session_instance_play_count_flow";

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/jig/codes/instance", port))
        .json(&serde_json::json!({
            "code": 123456,
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: PlayerSessionInstanceResponse = resp.json().await?;

    let token = body.token.clone();

    insta::assert_json_snapshot!(format!("{}-1",name), body, {".**.token" => "[instance_token]"});

    let resp = client
        .post(&format!(
            "http://0.0.0.0:{}/v1/jig/codes/instance/complete",
            port
        ))
        .json(&serde_json::json!({
            "token": token,
            "session": {
                "modules": [
                    {
                        "Matching": {
                            "stable_module_id": "00000000-0000-0000-0000-000000000000",
                            "rounds": [
                                {
                                    "2": {
                                        "failed_tries": 3
                                    },
                                    "0": {
                                        "failed_tries": 0
                                    },
                                    "3": {
                                        "failed_tries": 1
                                    }
                                },
                                {
                                    "1": {
                                        "failed_tries": 0
                                    },
                                    "2": {
                                        "failed_tries": 0
                                    },
                                    "0": {
                                        "failed_tries": 1
                                    }
                                }
                            ]
                        }
                    }
                ]
            }
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    Ok(())
}
