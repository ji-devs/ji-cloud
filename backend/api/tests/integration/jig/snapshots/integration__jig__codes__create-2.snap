---
source: tests/integration/jig/codes.rs
expression: body
---
{
  "codes": [
    {
      "index": "[index]",
      "jig_id": "0cc084bc-7c83-11eb-9f77-e3218dffb008",
      "name": null,
      "settings": {
        "direction": "ltr",
        "scoring": true,
        "dragAssist": true
      },
      "created_at": "[timestamp]",
      "expires_at": "[timestamp]"
    },
    {
      "index": "[index]",
      "jig_id": "0cc084bc-7c83-11eb-9f77-e3218dffb008",
      "name": null,
      "settings": {
        "direction": "ltr",
        "scoring": true,
        "dragAssist": true
      },
      "created_at": "[timestamp]",
      "expires_at": "[timestamp]"
    },
    {
      "index": "[index]",
      "jig_id": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
      "name": "test-name",
      "settings": {
        "direction": "rtl",
        "scoring": false,
        "dragAssist": false
      },
      "created_at": "[timestamp]",
      "expires_at": "[timestamp]"
    },
    {
      "index": "[index]",
      "jig_id": "3a71522a-cd77-11eb-8dc1-af3e35f7c743",
      "name": null,
      "settings": {
        "direction": "rtl",
        "scoring": true,
        "dragAssist": false
      },
      "created_at": "[timestamp]",
      "expires_at": "[timestamp]"
    }
  ]
}
